frame/support/src/lib.rs

```diff
@@ -946,3 +949,965 @@ mod tests {
+/// 	// Another declaration
+/// 	#[pallet::storage]
+/// 	#[pallet::getter(fn my_storage)]
+/// 	pub(super) type MyStorage<T> = StorageMap<_, Blake2_128Concat, u32, u32>;
+///
+/// 	// Declare genesis config. (This is optional)
+/// 	//
+/// 	// The macro accept either type alias or struct or enum, it checks generics are consistent.
+/// 	//
+/// 	// Type must implement `Default` traits
+/// 	#[pallet::genesis_config]
+/// 	#[derive(Default)]
+/// 	pub struct GenesisConfig {
+/// 		_myfield: u32,
+/// 	}
+///
+/// 	// Declare genesis builder. (This is need only if GenesisConfig is declared)
+/// 	#[pallet::genesis_build]
+/// 	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+/// 		fn build(&self) {}
+/// 	}
+///
+/// 	// Declare a pallet origin. (this is optional)
+/// 	//
+/// 	// The macro accept type alias or struct or enum, it checks generics are consistent.
+/// 	#[pallet::origin]
+/// 	pub struct Origin<T>(PhantomData<T>);
+///
+/// 	// Declare validate_unsigned implementation.
+/// 	#[pallet::validate_unsigned]
+/// 	impl<T: Config> ValidateUnsigned for Pallet<T> {
+/// 		type Call = Call<T>;
+/// 		fn validate_unsigned(
+/// 			source: TransactionSource,
+/// 			call: &Self::Call
+/// 		) -> TransactionValidity {
+/// 			Err(TransactionValidityError::Invalid(InvalidTransaction::Call))
+/// 		}
+/// 	}
+///
+/// 	// Declare inherent provider for pallet. (this is optional)
+/// 	//
+/// 	// The macro checks pallet is `Pallet<T>` or `Pallet<T, I>` and trait is `ProvideInherent`
+/// 	#[pallet::inherent]
+/// 	impl<T: Config> ProvideInherent for Pallet<T> {
+/// 		type Call = Call<T>;
+/// 		type Error = InherentError;
+///
+/// 		const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
+///
+/// 		fn create_inherent(_data: &InherentData) -> Option<Self::Call> {
+/// 			unimplemented!();
+/// 		}
+/// 	}
+///
+/// 	// Regular rust code needed for implementing ProvideInherent trait
+///
+/// 	#[derive(codec::Encode, sp_runtime::RuntimeDebug)]
+/// 	#[cfg_attr(feature = "std", derive(codec::Decode))]
+/// 	pub enum InherentError {
+/// 	}
+///
+/// 	impl sp_inherents::IsFatalError for InherentError {
+/// 		fn is_fatal_error(&self) -> bool {
+/// 			unimplemented!();
+/// 		}
+/// 	}
+///
+/// 	pub const INHERENT_IDENTIFIER: sp_inherents::InherentIdentifier = *b"testpall";
+/// }
+/// ```
+///
+/// # Example for pallet with instance.
+///
+/// ```
+/// #[frame_support::pallet]
+/// pub mod pallet {
+/// 	use frame_support::pallet_prelude::*;
+/// 	use frame_system::pallet_prelude::*;
+///
+/// 	type BalanceOf<T, I = ()> = <T as Config<I>>::Balance;
+///
+/// 	#[pallet::config]
+/// 	pub trait Config<I: 'static = ()>: frame_system::Config {
+/// 		#[pallet::constant]
+/// 		type MyGetParam: Get<u32>;
+/// 		type Balance: Parameter + From<u8>;
+/// 		type Event: From<Event<Self, I>> + IsType<<Self as frame_system::Config>::Event>;
+/// 	}
+///
+/// 	#[pallet::extra_constants]
+/// 	impl<T: Config<I>, I: 'static> Pallet<T, I> {
```

Comment:
On first reading this it's super confusing when `I` needs the `'static` bound and when it needs/allows for the default `= ()`.

Reply:
I agree, it probably needs a paragraph explaining instantiable modules.

* Basically all type must bound `= ()` so that pallet with default instance (this allow to not break when making a pallet previously non instantiable to instantiable).

* Then all impl block needs `'static` as it is needed in order to access `PalletInfo` implementation of for `Pallet`. Thus it is needed in order to use storages.

* Then some type allow to bound `'static` for the same reason it allows to bund `Config`, for convenience. (see Event allows to bound `Config`)

But the macro should show good error message when user forget some needed one.